src/crypto.rs

```diff
@@ -0,0 +1,158 @@
+        },
+        Algorithm::RS256 | Algorithm::RS384 | Algorithm::RS512 => {
+            let ring_alg = match algorithm {
+                Algorithm::RS256 => &signature::RSA_PKCS1_SHA256,
+                Algorithm::RS384 => &signature::RSA_PKCS1_SHA384,
+                Algorithm::RS512 => &signature::RSA_PKCS1_SHA512,
+                _ => unreachable!(),
+            };
+            // Taken from Ring doc https://briansmith.org/rustdoc/ring/signature/index.html
+            let key_pair = Arc::new(
+                signature::RSAKeyPair::from_der(
+                    untrusted::Input::from(key)
+                ).map_err(|_| ErrorKind::InvalidKey)?
+            );
+            let mut signing_state = signature::RSASigningState::new(key_pair)
+                .map_err(|_| ErrorKind::InvalidKey)?;
+            let mut signature = vec![0; signing_state.key_pair().public_modulus_len()];
+            let rng = rand::SystemRandom::new();
+            signing_state.sign(ring_alg, &rng, data.as_bytes(), &mut signature)
+                .map_err(|_| ErrorKind::InvalidKey)?;
+
+            Ok(base64::encode_config(
+                signature.as_ref(),
+                base64::URL_SAFE_NO_PAD
+            ))
+        },
+    }
+}
+
+/// Encode the claims passed and sign the payload using the algorithm from the header and the key
+pub fn encode<T: Serialize>(header: Header, claims: &T, key: &[u8]) -> Result<String> {
+    let encoded_header = to_jwt_part(&header)?;
+    let encoded_claims = to_jwt_part(&claims)?;
+    let payload = [encoded_header.as_ref(), encoded_claims.as_ref()].join(".");
+    let signature = sign(&*payload, key.as_ref(), header.alg)?;
+
+    Ok([payload, signature].join("."))
+}
+
+/// Compares the signature given with a re-computed signature for HMAC or using the public key (`key`)
+/// for RSA
+pub fn verify(signature: &str, payload: &str, key: &[u8], algorithm: Algorithm) -> Result<bool> {
+    match algorithm {
+        Algorithm::HS256 | Algorithm::HS384 | Algorithm::HS512 => {
+            // we just re-sign the data with the key and compare if they are equal
+            let signed = sign(payload, key, algorithm)?;
+            Ok(verify_slices_are_equal(signature.as_ref(), signed.as_ref()).is_ok())
+        },
+        Algorithm::RS256 | Algorithm::RS384 | Algorithm::RS512 => {
+            // we use ring to verify using the public key given
+            let verification_alg = match algorithm {
+                Algorithm::RS256 => &signature::RSA_PKCS1_2048_8192_SHA256,
+                Algorithm::RS384 => &signature::RSA_PKCS1_2048_8192_SHA384,
+                Algorithm::RS512 => &signature::RSA_PKCS1_2048_8192_SHA512,
+                _ => unreachable!(),
+            };
+            let res = signature::verify(
+                verification_alg,
+                untrusted::Input::from(key),
+                untrusted::Input::from(payload.as_bytes()),
+                untrusted::Input::from(signature.as_bytes()),
```

Comment:
Turns out it still doesn't work, despite being fed the same data as `pyjwt` after checking:

```rust
let signature_bytes = base64::decode_config(signature, base64::URL_SAFE_NO_PAD)?;
let res = signature::verify(
        verification_alg,
        untrusted::Input::from(key),
        untrusted::Input::from(signing_input.as_bytes()),
        untrusted::Input::from(signature_bytes.as_slice()),
);
// res is Err(Unspecified)
```

I'm using the same signature_bytes, key and signing_inpu below with `pyjwt` and the crypto library it uses: http://cryptography.io/

Equivalent in python of the RSA checking:
```py
from jwt.algorithms import get_default_algorithms
import base64

def base64url_decode(input):
    rem = len(input) % 4

    if rem > 0:
        input += b'=' * (4 - rem)

    return base64.urlsafe_b64decode(input)

public = open("public_rsa_key.pem", 'r').read()
encoded = b"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImprdSI6bnVsbCwia2lkIjpudWxsLCJ4NXUiOm51bGwsIng1dCI6bnVsbH0.eyJzdWIiOiJiQGIuY29tIiwiY29tcGFueSI6IkFDTUUifQ.a1P09WA30LTXyZV5KtOc2I_Ac8h0o47eVHdC_1S3JfarUs57q7sFoL-407aiFNrCR-QTU81bcnyYiJ-EJ8ZZXvdWvpgq6DU7_fEEnM19h4Phg0ZgKU3igv5XnkE9JXMn0Xk5DqUUQ1Nv4WyMjJM7zkVbYkkBik34MdCt39ldclJ9KRXLwwCsvaJTW8PGzv9e8lofE7Yl-d-11qU9pTYRD7T0o83rIPsEo_LmgY23galWbZmlqDpEXdQoTNT-gwee9V4M4S9jLJN8rPJgHlSFjDsaHcOdJDRowNOuPH1Oz2wbd3fhE3H9Sj82kNRBrxf-hkTQoIsZ9vELfD6395cYvg"
alg_obj = get_default_algorithms()["RS256"]
key = alg_obj.prepare_key(public)
signing_input, signature = encoded.rsplit(b".", 1)
print([i for i in signing_input])
print([i for i in base64url_decode(signature)])
print(alg_obj.verify(signing_input, key, base64url_decode(signature)))
```
The key is loaded differently in python (pem file) but the bytes are the same too.

I'll see if i can reproduce as a test case in `ring` and hoping it's not another silly mistake